### ISSUE TITLE
[HELP] Suspend Resume based on event

### ISSUE THREAD

**Issue by dust63** [elsa-workflows/elsa-core#1089]
I write an activity base on ObsWebSocket.

The role of this activity is to wait for a media ended event and resume workflow.
I don't really understood how to write that

For now I used a tack completion source but it's not really working

```
   private TaskCompletionSource _tcs;
        protected override async ValueTask<IActivityExecutionResult> OnExecuteAsync(ActivityExecutionContext context)
        {
            _ = base.OnExecuteAsync(context);          
            _tcs = new TaskCompletionSource();
            OBSWebsocket.MediaEnded += OnMediaEnded;
            await _tcs.Task;
            OBSWebsocket.MediaEnded -= OnMediaEnded;
            return Done(SourceEnded);
        }

        private void OnMediaEnded(OBSWebsocket sender, string sourceName, string sourceKind)
        {
            if (sourceName == SourceName || string.IsNullOrWhiteSpace(SourceName))
            {
                SourceEnded = sourceName;
                _tcs.SetResult();
            }        
        }
```

**Comment by sfmskywalker** [comment]
I will document this, but to give a couple of high-level pointers:

1. TCS won't do you any good for long-running workflows.
2. Instead, have your activity implement both `OnExecute` and `OnResume`. The former should return `Suspend()` and the latter might return `Done`.
3. Implement a bookmark class, e.g. `class ObsWebsocketBookmark : IBookmark`.
4. Impelement a bookmark provider, e.g. `class ObsWebsocketBookmarkProvider : IBookmark` -it only has to return an instance of `ObsWebsocketBookmark`.
5. Implement some background service thing or whatever is appropriate that listens for MediaEnded events. In that listener, use the `IWorkflowLaunchpad.CollectAndDispatchWorkflowsAsync`  or `IWorkflowLaunchpad.CollectAndExecuteWorkflowsAsync` methods.

The Telnyx activities use a similar model where some of the activities such as `Dial` continue only after having received a specific webhook event.

**Comment by dust63** [comment]
Thank you I will try

**Comment by sfmskywalker** [comment]
Here you go: https://elsa-workflows.github.io/elsa-core/docs/next/guides/guides-blocking-activities

Note: this is written against changes that are part of the master branch and to be released as 2.1. I.e. it won't compile against 2.0 packages.
